### DISCUSSION TITLE
Confusion around Sizedness bound for trait objects

### DISCUSSION THREAD

**Discussion by runiq** [pretzelhammer/rust-blog#50] [Q&A] [answered]
Hi, long-time reader, first-time poster :)

I got confused by two apparently conflicting bits in your whirlwind tour on Rust stdlib traits. In [the chapter on subtraits and supertraits](https://github.com/pretzelhammer/rust-blog/blob/master/posts/tour-of-rusts-standard-library-traits.md#subtraits--supertraits), you mention that only types can implement traits. Then, in [the chapter on sizedness](https://github.com/pretzelhammer/rust-blog/blob/master/posts/tour-of-rusts-standard-library-traits.md#subtraits--supertraits), you mention that all traits have an implicit `?Sized` bound like this:

```rust
trait Trait: ?Sized {}
```

Which, given that only types can implement traits according to the chapter on subtraits and supertraits, would be desugared into something like this:

```rust
trait Trait where Self: ?Sized {}
```

Now, to me that implies that wnenever I would implement *any* trait for a type, that type would automatically gain the `?Sized` bound. But that obviously cannot be true. So—what am I missing here? :thinking:

**Comment by pretzelhammer** [comment, marked as answer]
> Now, to me that implies that whenever I would implement any trait for a type, that type would automatically gain the `?Sized` bound. But that obviously cannot be true. So—what am I missing here? 🤔

I think your confusion might stem from how you're interpreting `?Sized`. So if `?Sized` meant _"is not sized"_ then yes. as an auto-bound on all trait impls it would not make sense since a lot of types are sized, however what it actually means is _"may not be sized"_ and any type may or may not be sized, just like any boolean may be true or may be false :)

In short: some types are `Sized`, some are types `!Sized`, and all types are `?Sized`, since `?Sized` is the superset of `Sized` + `!Sized`. Does that answer your question? If not feel free to ask follow-up questions :)

**Reply by runiq** [reply]
So basically, there's some "special sauce" for the 'relaxed bound' syntax that makes it behave somewhat differently from normal trait bounds?

If so, yes, that clears it up. Thanks! :)

**Reply by pretzelhammer** [reply]
Haha, you can call it "special sauce" if you want to but yes, a "relaxed bound" or a "widening bound" is different than a regular bound because it relaxes constraints instead of adding constraints, so it will always increase the number of types which can pass the constraints.
